From 0c82078e442c095899d2d3df16748db4b0c3aa48 Mon Sep 17 00:00:00 2001
From: qwang2 <quanyang.wang@windriver.com>
Date: Fri, 22 Jun 2018 13:16:24 +0800
Subject: [PATCH 6/8] regulator: core: free memory of rdev->dev.kobj.name to
 fix kmemleak

Kmemleak reports memory leak as below:

unreferenced object 0xa84abc40 (size 64):
comm "swapper/0", pid 1, jiffies 4294937386 (age 358.160s)
hex dump (first 32 bytes):
72 65 67 75 6c 61 74 6f 72 2e 36 00 ff 6b eb f2 regulator.6..k..
a1 4e d3 4b 08 ef 6d d1 d1 7e f2 bd ff fb 77 e7 .N.K..m..~....w.
backtrace:
	[<80a8a154>] kmemleak_alloc+0x40/0x74
	[<802655b8>] __kmalloc_track_caller+0x17c/0x2d0
	[<804e6120>] kvasprintf+0x50/0x9c
	[<804e61d4>] kvasprintf_const+0x68/0x70
	[<804dadc4>] kobject_set_name_vargs+0x34/0xa0
	[<805f771c>] dev_set_name+0x30/0x38
	[<8058db38>] regulator_register+0x388/0xc08
	[<8058f7e8>] devm_regulator_register+0x4c/0x84
	[<80591064>] anatop_regulator_probe+0x3d8/0x4f8
	[<805fe3d8>] platform_drv_probe+0x60/0xac
	[<805fc54c>] driver_probe_device+0x210/0x2d8
	[<805fc7bc>] __device_attach_driver+0xa4/0xbc
	[<805fa9e4>] bus_for_each_drv+0x98/0xa0
	[<805fc284>] __device_attach+0xac/0x114
	[<805fc7f0>] device_initial_probe+0x1c/0x20
	[<805fb81c>] bus_probe_device+0x38/0x90

In function regulator_register, dev_set_name is called to malloc a string
to assign to rdev->dev.kobj.name, so this string should be freed at the
fail path.

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/regulator/core.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/regulator/core.c b/drivers/regulator/core.c
index 87cf705..a7083c2 100644
--- a/drivers/regulator/core.c
+++ b/drivers/regulator/core.c
@@ -4429,6 +4429,7 @@ struct regulator_dev *
 	unset_regulator_supplies(rdev);
 	mutex_unlock(&regulator_list_mutex);
 wash:
+	kfree(rdev->dev.kobj.name);
 	kfree(rdev->constraints);
 	mutex_lock(&regulator_list_mutex);
 	regulator_ena_gpio_free(rdev);
-- 
1.7.9.5

